<!DOCTYPE html>
<html lang="en">

<!-- Head tag -->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">

  <!--Description-->
  

  <!--Author-->
  
  <meta name="author" content="Wade Zhao">
  

  <!--Open Graph Title-->
  
      <meta property="og:title" content="Using errors as control flow in Swift 译文"/>
  
  <!--Open Graph Description-->
  
  <!--Open Graph Site Name-->
  <meta property="og:site_name" content="Wade Zhao&#39;s Blog"/>
  <!--Type page-->
  
      <meta property="og:type" content="article" />
  
  <!--Page Cover-->
  

  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">

  <!-- Title -->
  
  <title>Using errors as control flow in Swift 译文 - Wade Zhao&#39;s Blog</title>


  <link rel="shortcut icon" href="https://hexo.io/icon/favicon-96x96.png">

  <!-- Custom CSS/Sass -->
  <link rel="stylesheet" href="/css/style.css">

  <!----------------------------
  https://github.com/GallenHu/hexo-theme-Daily

 _____            _   _
|  __ \          (_) | |
| |  | |   __ _   _  | |  _   _
| |  | |  / _` | | | | | | | | |
| |__| | | (_| | | | | | | |_| |
|_____/   \__,_| |_| |_|  \__, |
                          __/ |
                         |___/

    --------------------------->

</head>


<body>

  <!-- Nav -->
  <header class="site-header">
  <div class="header-inside">
    <div class="logo">
      <a href="/" rel="home">
        
        <img src="https://hexo.io/logo.svg" alt="Wade Zhao's Blog" height="60">
        
      </a>
    </div>
    <!-- Navigation -->
    <nav class="navbar">
      <!-- Collect the nav links, forms, and other content for toggling -->
      <div class="collapse">
        <ul class="navbar-nav">
          
          
            <li>
              <a href="/.">
                
                  home
                
              </a>
            </li>
          
            <li>
              <a href="/archives">
                
                  archive
                
              </a>
            </li>
          
        </ul>
      </div>
      <!-- /.navbar-collapse -->
    </nav>
    <div class="button-wrap">
      <button class="menu-toggle">Primary Menu</button>
    </div>
  </div>
</header>


  <!-- Main Content -->
  <div class="content-area">
  <div class="post">
    <!-- Post Content -->
    <div class="container">
      <article>
        <!-- Title date & tags -->
        <div class="post-header">
          <h1 class="entry-title">
            Using errors as control flow in Swift 译文
            
          </h1>
          <p class="posted-on">
          2019-01-15
          </p>
          <div class="tags-links">
            
              
                <a href="/tags/译文、Swift/" rel="tag">
                  译文、Swift
                </a>
              
            
          </div>
        </div>
        <!-- Post Main Content -->
        <div class="entry-content has_line_number">
          <h1 id="Using-errors-as-control-flow-in-Swift"><a href="#Using-errors-as-control-flow-in-Swift" class="headerlink" title="Using errors as control flow in Swift"></a>Using errors as control flow in Swift</h1><p>我们如何对APP和系统中的控制流进行管理，会在代码执行速度和调试难易程度上产生全方面的影响。我们代码的控制流本质上是执行各种函数和语句的顺序，以及最终输入的代码路径。</p>
<p>Swift 提供多种多样的工具来定义控制流-例如<figure class="highlight plain"><figcaption><span>或者类似于可选值的结构。这周，就让我们来看看怎么构建Siwft的错误抛出和构建模型来确保我们的控制流能够更容易地去管理。</span></figcaption><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">## 抛弃可选类型</span><br><span class="line">可选值，是一个非常重要的语言特性和一个很棒的方式去模型化那些能够被合法忽略的数据，在涉及给定函数内的控制流时，他通常可以被当做一个模板去使用。</span><br><span class="line"></span><br><span class="line">在这里我们有一个从APP Bundle中加载一张图片的函数，接下来对这张图片tint和重新调整大小。由于每个操作都在当前会返回一个可选类型的图片，我们最终会有保护我们函数退出的语句。</span><br></pre></td></tr></table></figure></p>
<p>func loadImage(named name: String,<br>               tintedWith color: UIColor,<br>               resizedTo size: CGSize) -&gt; UIImage? {<br>    guard let baseImage = UIImage(named: name) else {<br>        return nil<br>    }</p>
<pre><code>guard let tintedImage = tint(baseImage, with: color) else {
    return nil
}

return resize(tintedImage, to: size)
</code></pre><p>}<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">我们这样处理会面临的问题是我们基本上都会使用nil值来处理运行时的错误-这样我们每次都要进行解包，并且会从第一处隐藏错误发生的潜在原因。</span><br><span class="line"></span><br><span class="line">让我们看看如何使用抛出函数和错误来解决这两个问题。我们首先定义一个在处理图像代码中可能出现的每个错误的枚举 - 看起来像这样：</span><br></pre></td></tr></table></figure></p>
<p>enum ImageError: Error {<br>    case missing<br>    case failedToCreateContext<br>    case failedToRenderImage<br>    …<br>}<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">我们将更改所有内部函数，以便在失败时抛出上述错误之一，而不是返回nil。 例如，以下是我们如何快速更新loadImage（命名:)以返回非可选的UIImage或抛出ImageError.missing：</span><br></pre></td></tr></table></figure></p>
<p>private func loadImage(named name: String) throws -&gt; UIImage {<br>    guard let image = UIImage(named: name) else {<br>        throw ImageError.missing<br>    }</p>
<pre><code>return image
</code></pre><p>}<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">一旦我们对其他图像处理函数做相同的处理，我们一样可以将相同的更改应用到我们的顶级函数 重，删除所有可选值并使其返回具体图像或抛出在我们的函数中产生的任何错误操作：</span><br></pre></td></tr></table></figure></p>
<p>func loadImage(named name: String,<br>               tintedWith color: UIColor,<br>               resizedTo size: CGSize) throws -&gt; UIImage {<br>    var image = try loadImage(named: name)<br>    image = try tint(image, with: color)<br>    return try resize(image, to: size)<br>}<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">上述更改不仅会使我们的函数体更简单 - 它还使得调试更加容易，因为我们现在在出现问题时最后会出现定义好的错误，而不是需要弄清楚导致返回nil的原因是什么。</span><br><span class="line"></span><br><span class="line">但是我们在实际处理中并不是对所有错误感兴趣。所以我们不希望在我们的代码库中的任何地方都要求使用do，try，catch模式（具有讽刺意味的是，试图避免产生大多数相同的模板，而不是一味地用调用一套模板）。</span><br><span class="line"></span><br><span class="line">好消息是，我们可以随时返回optional - 即使使用了throw来处理错误。难道我们所要做的就是使用try？ 当调用throw函数时的关键字，我们也可以获得一个可选值的返回值：</span><br></pre></td></tr></table></figure></p>
<p>let optionalImage = try? loadImage(<br>    named: “Decoration”,<br>    tintedWith: .brandColor,<br>    resizedTo: decorationSize<br>)<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">`try`有什么好处？它是一个两全其美的解决方案。我们能够在获得一个可选值的同时仍让我们利用`throw`和`errors`的力量来管理我们的内部控制流程</span><br><span class="line"></span><br><span class="line">## 验证输入</span><br><span class="line"></span><br><span class="line">接下来，让我们看一下在输入验证的场景下如何使用`errors`改进控制流。即使Swift有一个非常先进和强大的类型系统，它也不能总是确保我们的函数将接收有效的输入，所以运行时检查有时是我们唯一的选择。</span><br><span class="line"></span><br><span class="line">让我们看一下另一个例子，当我们在注册新帐户时需要去验证用户选择的凭据。就像以前一样，当前的代码对每个验证规则都使用保护语句，并在发生故障时输出错误消息 - 如下所示：</span><br></pre></td></tr></table></figure></p>
<p>func signUpIfPossible(with credentials: Credentials) {<br>    guard credentials.username.count &gt;= 3 else {<br>        errorLabel.text = “Username must contain min 3 characters”<br>        return<br>    }</p>
<pre><code>guard credentials.password.count &gt;= 7 else {
    errorLabel.text = &quot;Password must contain min 7 characters&quot;
    return
}

// Additional validation
...

service.signUp(with: credentials) { result in
    ...
}
</code></pre><p>}<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">即使我们只验证上面的两个数据，我们验证逻辑的代码数量的增长速度将会超过我们的预期。将这种逻辑与我们的UI代码（通常在`ViewController`中）一起使用也会使测试变得更加困难。所以让我们看看是否可以进行一些解耦并改善验证逻辑的控制流程。</span><br><span class="line"></span><br><span class="line">理想情况下，我们希望我们能够自己包含验证代码。这样，它既可以单独处理也可以单独测试，也可以轻松重用在我们的其他代码中。为了实现这一点，让我们首先为所有验证逻辑创建一个专用类型。 我们将其称为Validator，并使其成为一个简单的结构，它包含给定Value类型的验证闭包：</span><br></pre></td></tr></table></figure></p>
<p>struct Validator<value> {<br>    let closure: (Value) throws -&gt; Void<br>}<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">使用上面的代码，我们将能够构造验证器，只要值没有通过验证就会抛出错误。 但是，必须始终为每个验证过程定义一个新的错误类型，在这个过程中可能会再次生成不必要的模板（特别是如果我们想要把错误显示给用户）- 这样我们可以编写验证的函数，简单地传递一个Bool值和一条消息，以便在发生故障时显示给用户：</span><br></pre></td></tr></table></figure></value></p>
<p>struct ValidationError: LocalizedError {<br>    let message: String<br>    var errorDescription: String? { return message }<br>}</p>
<p>func validate(<br>    _ condition: @autoclosure () -&gt; Bool,<br>    errorMessage messageExpression: @autoclosure () -&gt; String<br>) throws {<br>    guard condition() else {<br>        let message = messageExpression()<br>        throw ValidationError(message: message)<br>    }<br>}<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">上面我们再次使用@autoclosure，它允许我们通过自动将表达式解包在闭包中来推迟表达式。有关更多信息，请查看[Using @autoclosure when designing Swift APIs](https://www.swiftbysundell.com/posts/using-autoclosure-when-designing-swift-apis)</span><br><span class="line"></span><br><span class="line">有了上述内容，我们现在可以将所有验证逻辑实现为专用验证器 - 使用Validator类型上的计算静态属性构建。 例如，以下是我们如何为密码实现验证器：</span><br></pre></td></tr></table></figure></p>
<p>extension Validator where Value == String {<br>    static var password: Validator {<br>        return Validator { string in<br>            try validate(<br>                string.count &gt;= 7,<br>                errorMessage: “Password must contain min 7 characters”<br>            )</p>
<pre><code>        try validate(
            string.lowercased() != string,
            errorMessage: &quot;Password must contain an uppercased character&quot;
        )

        try validate(
            string.uppercased() != string,
            errorMessage: &quot;Password must contain a lowercased character&quot;
        )
    }
}
</code></pre><p>}<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">为了解决这个问题，让我们再次重载一个```Validate```作为一个语法糖，通过调用要验证的值和使用的验证器。</span><br></pre></td></tr></table></figure></p>
<p>func validate<t>(_ value: T,<br>                 using validator: Validator<t>) throws {<br>    try validator.closure(value)<br>}<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">有了所有构建block，让我们更新调用的地方来使用我们的新验证系统。上述方法的优点在于，虽然需要一些额外的类型和一点点基础设施，但它让我们的代码在需要输入验证非常的clean：</span><br></pre></td></tr></table></figure></t></t></p>
<p>func signUpIfPossible(with credentials: Credentials) throws {<br>    try validate(credentials.username, using: .username)<br>    try validate(credentials.password, using: .password)</p>
<pre><code>service.signUp(with: credentials) { result in
    ...
}
</code></pre><p>}<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">也许更好的是，我们现在可以通过使用do，try，catch的模式和调用上面的signUpIfPossible函数来处理所有验证错误，然后简单地向用户显示任何抛出错误的本地化描述：</span><br></pre></td></tr></table></figure></p>
<p>do {<br>    try signUpIfPossible(with: credentials)<br>} catch {<br>    errorLabel.text = error.localizedDescription<br>}<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">值得注意的是，虽然上面的代码示例没有使用任何本地化，但我们总是希望在真实应用程序中向用户显示所有错误消息时使用本地化字符串。</span><br><span class="line"></span><br><span class="line">## 抛出测试(Throwing Tests)</span><br><span class="line">围绕可能遇到的错误来构造代码的另一大好处是，它通常使测试更容易。由于抛出函数本质上具有两个不同的可能输出-正确的值和错误 - 添加覆盖这两种情形的测试在很多情况下非常简单。</span><br><span class="line"></span><br><span class="line">例如，以下是我们如何轻松地为我们的密码验证代码添加测试 - 通过简单地断言到一个错误情况来进行涵盖我们的两个要求，并且成功案例不会：</span><br></pre></td></tr></table></figure></p>
<p>class PasswordValidatorTests: XCTestCase {<br>    func testLengthRequirement() throws {<br>        XCTAssertThrowsError(try validate(“aBc”, using: .password))<br>        try validate(“aBcDeFg”, using: .password)<br>    }</p>
<pre><code>func testUppercasedCharacterRequirement() throws {
    XCTAssertThrowsError(try validate(&quot;abcdefg&quot;, using: .password))
    try validate(&quot;Abcdefg&quot;, using: .password)
}
</code></pre><p>}<br><code>`</code></p>
<p>正如您在上面所看到的，因为XCTest支持<code>throw</code>测试函数 - 并且每个未处理的错误都算作失败 - 所有我们要做的就是使用try调用我们的validate函数来验证成功的情况，如果函数没有抛出错误，那我们的测试就是成功👍。</p>
<h2 id="总结"><a href="#总结" class="headerlink" title="总结"></a>总结</h2><p>虽然有许多方法可以组织Swift代码的控制流 - 对于可能成功或失败的操作，使用错误和<code>throw</code>可能是一个很好的选择。 虽然这样做需要一些额外的代码（例如引入错误类型，并且确保所有的调用都是<code>try</code>或者<code>try?</code>） - 它可以给我们一些实际的好处，同时也使我们的代码更紧凑。</p>
<p>从一些函数中返回可选值也是合适的 - 特别是那些没有任何明显错误的函数 - 但是在我们处理几个不同的可选值和<code>guard</code>语句的地方 - 使用错误代替可能会给我们一个 更加清晰的控制流程。</p>
<p>你怎么看？您目前是否使用错误和<code>throw</code>来管理代码的控制流程 - 或者您将尝试使用它？ 请在Twitter @johnsundell上告诉我 - 以及您的问题，意见和反馈。</p>
<p>感谢阅读！<br>文章中有翻译不周的地方，可以联系我微博：<a href="https://weibo.com/u/5300292054?wvr=5&amp;is_all=1#_rnd1551097501695" target="_blank" rel="noopener">我会每天早睡的</a></p>

        </div>
      </article>
    </div>
    <!-- Comments -->
    <div class="container">
      
<section id="comment">
  <!-- <h1 class="title">Comments</h1> -->

  
  <div id="disqus_thread">
    <script type="text/javascript">
    var disqus_config = function () {
          this.page.url = window.location.href;
          this.page.identifier = 'using-译文';
          this.page.title = 'Using errors as control flow in Swift 译文';
      };
    </script>
    <noscript>Please enable JavaScript to view the <a href="//disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
  </div>
  
</section>


    </div>
    <!-- Pre or Next -->
    <div class="nav-links">
      
      
        <div class="nav-next">
          <a href="/2018/11/08/并行解决诡异问题/" rel="prev">Newer Posts <span class="meta-arraw meta-arraw-right"></span></a>
        </div>
      
    </div>

  </div>
</div>


  <!-- Footer -->
  <!-- Footer-widgets -->
<div class="footer-widgets">
  <div class="row inside-wrapper">
    <div class="col-1-3">
      <aside>
        <h1 class="widget-title">关于本站</h1>
        <div class="custom-widget-content">
          
          <p>These violent delights have violent ends ,and in their triumph die ,like fire and powder ,which,as they kiss ,consume</p>
        </div>
      </aside>
    </div>
    <div class="col-1-3">
      <aside>
        <h1 class="widget-title">Contact</h1>
        <div class="widget-text">
          
            
              <a href="https://github.com/WadeZhao23" class="icon icon-github" target="_blank">github</a>
            
              <a href="https://weibo.com/5300292054/profile?rightmod=1&wvr=6&mod=personinfo" class="icon icon-weibo" target="_blank">weibo</a>
            
              <a href="zhaoyuwei23@gmail.com" class="icon icon-mail" target="_blank">mail</a>
            
              <a href="https://twitter.com/Tweg24" class="icon icon-twitter" target="_blank">twitter</a>
            
          
        </div>
      </aside>
    </div>
    <div class="col-1-3">
      <aside>
        <h1 class="widget-title">Search</h1>
        <div class="widget-text">
          <form onSubmit="return appDaily.submitSearch('')">
            <p>
              <input type="text" placeholder="search..." id="homeSearchInput">
            </p>
            <!-- <input type="submit" value="GO"> -->
          </form>
        </div>
      </aside>
    </div>
  </div>
</div>
<!-- Footer -->
<footer class="site-info">
  <p>
    <span>Wade Zhao's Blog &copy; 2019</span>
    
      <span class="split">|</span>
      <span>Powered by <a href="https://hexo.io/" target="_blank">Hexo</a> with Theme <a href="https://github.com/GallenHu/hexo-theme-Daily" target="_blank">Daily</a></span>
    
  </p>
</footer>


  <!-- After footer scripts -->
  <!-- scripts -->
<script src="/js/app.js"></script>

<script>
  var disqus_shortname = 'WadeZhao';

  
  var disqus_url = 'http://yoursite.com/2019/01/15/译文/';
  

  (function(){
    var dsq = document.createElement('script');
    dsq.type = 'text/javascript';
    dsq.async = true;
    dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
    (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
  })();
</script>





</body>

</html>
